File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.cpp
class Solution {
public:
    int backtrack(vector<int>& nums, int index, int prevIndex,
                  vector<vector<int>>& dp) {

        if (index == nums.size())
            return 0;

        if (dp[index][prevIndex + 1] != -1)
            return dp[index][prevIndex + 1];

        int skip = backtrack(nums, index + 1, prevIndex, dp);

        int take = 0;

        if (prevIndex == -1 || nums[index] > nums[prevIndex])
            take = 1 + backtrack(nums, index + 1, index, dp);

        return dp[index][prevIndex + 1] = max(take, skip);
    }

    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(n + 1, -1));

        return backtrack(nums, 0, -1, dp);
    }
};

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-2.java
class Solution {

    boolean solve(int index, int target, int[] nums, int[][] dp) {

        if (index == nums.length) {
            return target == 0;
        }

        if (dp[index][target] != -1) {
            return dp[index][target] == 1;
        }

        boolean take = false;

        if (nums[index] <= target) {
            take = solve(index + 1, target - nums[index], nums, dp);
        }

        boolean skip = solve(index + 1, target, nums, dp);

        dp[index][target] = (take || skip) ? 1 : 0;

        return take || skip;
    }

    public boolean canPartition(int[] nums) {

        int sum = 0;

        for (int n : nums)
            sum += n;

        if (sum % 2 == 1)
            return false;

        int target = sum / 2;

        int[][] dp = new int[nums.length][target + 1];

        for (int i = 0; i < nums.length; i++) {
            Arrays.fill(dp[i], -1);
        }

        return solve(0, target, nums, dp);
    }
}